src/v/cloud_topics/core/write_pipeline.h

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#pragma once
+
+#include "cloud_topics/core/event_filter.h"
+#include "cloud_topics/core/pipeline_stage.h"
+#include "cloud_topics/core/write_request.h"
+#include "ssx/semaphore.h"
+
+#include <seastar/core/abort_source.hh>
+#include <seastar/core/condition-variable.hh>
+#include <seastar/core/gate.hh>
+#include <seastar/core/loop.hh>
+#include <seastar/core/lowres_clock.hh>
+
+#include <exception>
+#include <type_traits>
+
+namespace experimental::cloud_topics::core {
+
+struct write_request_process_result {
+    /// Iteration should be stopped
+    ss::stop_iteration stop_iteration{ss::stop_iteration::yes};
+    /// Current write request is fully processed and should be moved to the next
+    /// staged
+    bool advance_next_stage{false};
+};
+
+struct write_pipeline_accessor;
+
+template<class Clock = ss::lowres_clock>
+class write_pipeline {
```

Comment:
missing stop() method and a vassert(_gate.is_closed(), ...) in destructor?

Reply:
it's in the next PR

Reply:
what is "next PR"?